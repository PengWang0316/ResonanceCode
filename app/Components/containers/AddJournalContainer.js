import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import JournalForm from '../JournalForm';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';
import { checkAuthentication } from '../../actions/UserActions';
import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import { updateJournal, createJournal, deleteJournal, deleteUnattachedJournal } from '../../actions/JournalActions';
import { clearReadings } from '../../actions/ReadingActions';

/**
 * Add Journal Container class.
 */
class AddJournalContainer extends Component {
  /** Checking authentication if user object cannot be found in props.
   * @returns {null} No return.
   */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    this.isSubmited = false; // Use this as an indicator for finishing submit.
  }

  /** After updating, creating and deleting, forward to reading page.
    * @param {object} nextProps is the object that contains new props.
    * @returns {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if (this.isSubmited && !this.props.journal && this.props.isLoading && !nextProps.isLoading) {
      this.props.clearReadings();
      this.props.history.push('/reading');
    }
  }

  /*
    ** updateObject = {
        isUpdate: bool,
        journalId: ,
        journalDate: ,
        userId: ,
        readings: ,
        contents: ,
        oldReadingIds:
      }
    */

  /** Handling the submit event.
   * @param {object} updateObject is the journal object will be updated or created.
   * @returns {null} No return.
   */
  handleSubmitCallback = updateObject => {
    // this.props.isLoading(true);
    const journal = Object.assign({
      _id: updateObject.journalId,
      date: new Date(updateObject.journalDate),
      // user_id: updateObject.userId,
      readings: updateObject.readings,
      shareList: updateObject.shareList
    }, updateObject.contents);
    // console.log("submit:", updateObject);
    if (updateObject.isUpdate) {
      // console.log("oldContentKeys:", this.oldContentKeys);
      // console.log("oldReadingKeys:", this.oldReadingIds);
      // Assemble two arrays for deletion contents and reading
      // const deleteContents = [];
      const deleteReadingIds = [];
      // Object.prototype.hasOwnProperty.call(updateObject.readings) this.oldContentKeys.map((element)=>{if(!this.contents.hasOwnProperty(element)) deleteContents.push(element);});
      // if this.oldReadingIds array is null, it is a unattached journal
      /* if (updateObject.oldReadingIds) updateObject.oldReadingIds.map((element) => { if (!updateObject.readings.hasOwnProperty(element)) deleteReadingIds.push(element); }); */
      if (updateObject.oldReadingIds) updateObject.oldReadingIds
        .forEach((element) => {
          if (!Object.prototype.hasOwnProperty.call(updateObject.readings, element))
            deleteReadingIds.push(element);
        });

      // journal.deleteContents = deleteContents; //todo delete deleteContents and oldContentKeys
      journal.deleteReadingIds = deleteReadingIds;
      journal.isUnattachedJournal = !updateObject.oldReadingIds;
      // console.log("delete contents: ",deleteContents);
      // console.log("delete reading ids : ",deleteReadingIds);
      // updateJournal(journal, !updateObject.oldReadingIds).then((result) => {
      //   this.removeLoadingAndForward();
      // });
      this.props.updateJournal(journal);
    } else this.props.createJournal(journal);
    // this.removeLoadingAndForward();
    // createJournal(journal).then((result) => {
    //   this.removeLoadingAndForward();
    // });
    this.isSubmited = true;
  }

  /** Handling the delete event.
   * @param {string} journalId is the id for the journal.
   * @param {string} readingIds is a reading's id array that contains all reading connects to this journal.
   * @param {boolean} isUnattachedJournal is a indicator of whether is a unattached journal.
   * @returns {null} No return.
   */
  handleDeleteCallback = (journalId, readingIds, isUnattachedJournal) => {
    // console.log("Delete journal!");
    // event.preventDefault();
    // this.props.isLoading(true);
    this.props.clearJournalState();
    if (isUnattachedJournal) {
      this.props.deleteUnattachedJournal(journalId);
      this.props.history.push('/reading');
    } else this.props.deleteJournal({ journalId, readingIds });
    // deleteJournal(journalId, readingIds, userId).then(reault => {
    //   this.removeLoadingAndForward();
    // });
  }

  /*
  removeLoadingAndForward() {
    // this.props.isLoading(false);
    this.props.clearJournalState();
    this.props.history.push('/reading');
  }
*/
  /** Rendering the component.
   * @returns {jsx} Return the jsx that represents the component.
   */
  render() {
    // console.log("container: ", this.props.journal);
    return (
      <UnauthenticatedUserCheck>
        <div>
          <LoadingAnimation />
          <JournalForm
            journalData={this.props.journal}
            isWriting={this.props.isLoading}
            history={this.props.history}
            handleSubmit={this.handleSubmitCallback}
            handleDelete={this.handleDeleteCallback}
          />
        </div>
      </UnauthenticatedUserCheck>
    );
  }
}

/*
AddJournalContainer.propTypes = {
  journalData: PropTypes.object,
  history: PropTypes.object
}; */

const mapStateToProps = state => ({
  isLoading: state.isLoading,
  journal: state.journal,
  user: state.user
});

const mapDispatchToProps = dispatch => ({
  // isLoading: loadingState => dispatch(isLoading(loadingState)),
  checkAuthentication: _ => dispatch(checkAuthentication()),
  createJournal: journal => dispatch(createJournal(journal)),
  updateJournal: journal => dispatch(updateJournal(journal)),
  deleteJournal: params => dispatch(deleteJournal(params)),
  deleteUnattachedJournal: journalId => dispatch(deleteUnattachedJournal(journalId)),
  clearReadings: _ => dispatch(clearReadings())
});
export default connect(mapStateToProps, mapDispatchToProps)(AddJournalContainer);
