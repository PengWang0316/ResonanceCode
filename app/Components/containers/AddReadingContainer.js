import React, { Component } from 'react';
import { connect } from 'react-redux';
import jQuery from 'jquery';
import { JWT_MESSAGE } from '../../config';
import AddReadingForm from '../AddReadingForm';
// import Loading from '../Loading';
// import LoginApi from '../../apis/LoginApi';
import { getImageArray } from '../../apis/Util';
// import { createReading } from '../../apis/DatabaseApi';
import { checkAuthentication } from '../../actions/UserActions';

import { clickCoin, clearAddReadingTempState, createReading } from '../../actions/ReadingActions';
// import isLoading from '../../actions/LoadingActions';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';

/**
 * Add reading container
 * @returns {null} No return
 */
class AddReadingContainer extends Component {
  /**
  * Eliminating empty string from changeLines.
  * @param {string} changeLines is current change lines' string.
  * @returns {string} Return a new string without empty string.
  */
  static eliminateEmptyString(changeLines) {
    return changeLines.split(' ').filter((element) => element.length > 0).join(',');
  }

  /**
   * Handle subcomponent's cancel callback.
   * @returns {null} No return.
   */
  static handleCancelCallback = () =>
    $('#addReadingModal').modal('toggle'); // This method will use the $ that comes from index.html page.

  /**
   * Checking authentication
   * @returns {null} No return
   */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    this.props.clearAddReadingTempState();
  }

  /**
   * Clearing addReadingTempState, unlocking submit buttons, and turn off the addReadingForm (modal).
   * @param {nextProps} nextProps is an object that contains new props.
   * @returns {null} No return
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.readings.length !== 0 &&
      this.props.readings.length < nextProps.readings.length) {
      AddReadingContainer.handleCancelCallback();
      this.props.clearAddReadingTempState();
      jQuery('button[type=submit]').attr('disabled', '');
    }
  }

  /**
   * Handle subcomponent's coin click callback.
   * @param {int} lineNumber is the number of line.
   * @param {int} coinsPoint is the point that moment this line has based on the calculation.
   * @returns {null} No return.
   */
  handleCoinClickCallback = (lineNumber, coinsPoint) => {
    // console.log("LineNumber:",lineNumber);
    // console.log("coin point:",coinsPoint);
    const changeLinesArr = ['1st', '2nd', '3rd', '4th', '5th', '6th'];
    const regExpArr = [/1st/, /2nd/, /3rd/, /4th/, /5th/, /6th/];
    // change the state based on result
    const addReadingTempState = Object.assign({}, this.props.addReadingTempState); /* Making a copy from previous state */
    // console.log("Previous state: ", addReadingTempState);
    let side1;
    let middle1;
    let side2;
    let middle2;
    const classSide = 'img-line-side-big';
    const classMiddle = 'img-line-middle-big';
    const classMiddleB = 'img-line-middle-blank-big';
    // const classRed = 'background-red-big';
    const classRedSide = 'img-line-side-red-big';
    const classRedMiddle = 'img-line-middle-red-big';
    // console.log("b:",this.state.changeLines);
    // console.log("r:",this.changeLinesArr[lineNumber]);
    addReadingTempState.changeLines = addReadingTempState.changeLines.replace(regExpArr[lineNumber], '');
    const changeLinesIndex = addReadingTempState.changeLinesNumberArray.indexOf(`${lineNumber}`);
    if (changeLinesIndex !== -1) addReadingTempState.changeLinesNumberArray
      .splice(changeLinesIndex, 1);

    // console.log("cda:",this.state.changeLines);
    switch (coinsPoint) {
      case 6:
        side1 = classRedSide;
        side2 = classRedSide;
        middle1 = classMiddleB;
        middle2 = classRedMiddle;
        // erase the word if it has already exsited
        // console.log("changeLinesArr",this.changeLinesArr[lineNumber]);
        // this.state.changelines=this.state.changeLines.replace(this.changeLinesArr[lineNumber],"");
        addReadingTempState.changeLines += addReadingTempState.changeLines.length === 0 ? changeLinesArr[lineNumber] : `  ${changeLinesArr[lineNumber]}`;
        addReadingTempState.changeLinesNumberArray.push(lineNumber);// Also put the line number in an array
        // put number to image array
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = 7;
        break;
      case 7:
        side1 = classSide;
        middle1 = classMiddle;
        side2 = classSide;
        middle2 = classMiddle;
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = coinsPoint;
        break;
      case 9:
        side1 = classRedSide;
        side2 = classRedSide;
        middle1 = classRedMiddle;
        middle2 = classMiddleB;
        // erase the word if it has already exsited
        // this.state.changelines=this.state.changeLines.replace(this.changeLinesArr[lineNumber],"");
        addReadingTempState.changeLines += addReadingTempState.changeLines.length === 0 ? changeLinesArr[lineNumber] : `  ${changeLinesArr[lineNumber]}`;
        addReadingTempState.changeLinesNumberArray.push(lineNumber);// Also put the line number in an array
        // put number to image array
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = 8;
        break;
      case 8:
        side1 = classSide;
        side2 = classSide;
        middle1 = classMiddleB;
        middle2 = classMiddleB;
        addReadingTempState.imageArrays.img1[lineNumber] = coinsPoint;
        addReadingTempState.imageArrays.img2[lineNumber] = coinsPoint;
        break;
      default:
        break;
    }
    // console.log(`line${lineNumber}`);
    addReadingTempState[`line${lineNumber}`].side1 = side1;
    addReadingTempState[`line${lineNumber}`].side2 = side2;
    addReadingTempState[`line${lineNumber}`].middle1 = middle1;
    addReadingTempState[`line${lineNumber}`].middle2 = middle2;

    // set next line to be available
    addReadingTempState.availableArr[(lineNumber * 1) + 1] = true;
    // console.log(this.eliminateEmptyString(this.state.changeLines));
    // console.log(this.imageArrays);
    this.props.clickCoin(addReadingTempState); /* Make sure addReadingTempState is a new object */
  }

  /**
   * Handle subcomponent's submit callback.
   * @param {object} readingObject is the new reading object.
   * @returns {null} No return.
   */
  handleSubmitCallback(readingObject) {
    // if (event) event.preventDefault();
    // console.log("Submit");
    // locking screen and call api
    // this.setState({isWriting:true});
    // this.props.loading(true);
    const img1 = this.props.addReadingTempState.imageArrays.img1.join();
    const img2 = this.props.addReadingTempState.imageArrays.img2.join();

    const reading = {
      reading_name: readingObject.readingName,
      hexagram_arr_1: getImageArray(img1),
      hexagram_arr_2: getImageArray(img2),
      img1,
      img2,
      date: new Date(readingObject.date), // keeping a date object to mongoDB
      change_lines: this.props.addReadingTempState.changeLinesNumberArray,
      change_lines_text: AddReadingContainer
        .eliminateEmptyString(this.props.addReadingTempState.changeLines),
      people: readingObject.people,
      // user_id: LoginApi.isLogin(document).userid
    };
    // Before send to the server, setting up the right local time
    const currentTime = new Date();
    reading.date.setHours(currentTime.getHours());
    reading.date.setMinutes(currentTime.getMinutes());
    reading.date.setSeconds(currentTime.getSeconds());
    // Calling the action.
    this.props.createReading({ reading, jwtMessage: localStorage.getItem(JWT_MESSAGE) });
    /* Deprecated old version
    createReading(reading).then((result) => {
      // console.log("result:",result);
      this.props.loading(false);
      this.props.history.push('/reading');
    }); */
  }

  /**
   * The render method for this component.
   * @returns {null} No return.
   */
  render() {
    return (
      <div className="modal fade" id="addReadingModal" tabIndex="-1" role="dialog" aria-labelledby="addReadingModalLabel" aria-hidden="true">
        <div className="modal-dialog modal-lg" role="document">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title" id="addReadingModalLabel">Add a new reading</h5>
              <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                <span aria-hidden="true">&times;</span>
              </button>
            </div>
            <div className="modal-body">
              <LoadingAnimation />
              <AddReadingForm
                handleSubmit={(readingObject) => this.handleSubmitCallback(readingObject)}
                handleCancel={AddReadingContainer.handleCancelCallback}
                handleCoinClick={this.handleCoinClickCallback}
                addReadingTempState={this.props.addReadingTempState}
              />
            </div>
          </div>
        </div>
      </div>

    );
  }
}

const mapStateToProps = state => ({
  addReadingTempState: state.addReadingTempState,
  // isLoadingState: state.isLoading,
  user: state.user,
  readings: state.readings
});
const mapDispatchToProps = dispatch => ({
  clickCoin: addReadingTempState => dispatch(clickCoin(addReadingTempState)),
  clearAddReadingTempState: _ => dispatch(clearAddReadingTempState()),
  // loading: isLoadingBool => dispatch(isLoading(isLoadingBool)),
  checkAuthentication: _ => dispatch(checkAuthentication()),
  createReading: params => dispatch(createReading(params))
});
export default connect(mapStateToProps, mapDispatchToProps)(AddReadingContainer);
