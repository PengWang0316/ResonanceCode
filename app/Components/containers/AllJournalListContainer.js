import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import styles from '../../styles/AllJournalListContainer.module.css';
import LoadingAnimation from '../SharedComponents/LoadingAnimation';
import JournalRow from '../JournalRow';
import Pagination from '../SharedComponents/Pagination';
import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import { checkAuthentication } from '../../actions/UserActions';
import { fetchAllJournal } from '../../actions/JournalActions';
import { NUMBER_PER_PAGE_JOURNAL } from '../../config';

/** Show all journal a user has. */
export class AllJournalListContainer extends Component {
  state = {
    journals: []
  };
  /** Checking the users' authentication state and calling fetchAllJournal function from JournalActions.
    * @return {null} No return.
  */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
    else if (this.props.allJournal.length === 0) this.props.fetchAllJournal();
    // When the component is mounted second time, all props will remind the same. The componentWillReceivedProps method will not be called. (Not sure whether it is a bug.) So, the code below can add a checking to prevent this happen.
    if (this.state.journals.length === 0 && this.props.allJournal.length !== 0)
      this.setState({ journals: this.props.allJournal.slice(0, NUMBER_PER_PAGE_JOURNAL) });
    // this.currentPage = 0;
  }

  /** Calling fetchAllJournal function from JournalActions after the user login.
    * Setting the first showing journal to state.
    * @param {object} nextProps is an object that contains props' vaule.
    * @return {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if (!this.props.user.isAuth && nextProps.user.isAuth && this.props.allJournal.length === 0)
      this.props.fetchAllJournal();
    if (this.state.journals.length === 0 && nextProps.allJournal.length !== 0)
      this.setState({ journals: nextProps.allJournal.slice(0, NUMBER_PER_PAGE_JOURNAL) });
  }

  /** Getting the part of content from props.allJournal state and setting them to the component state.
    * @param {number} currentPage is the number of the page a user is clicking.
    * @return {null} No return.
  */
  changeContent = currentPage => {
    this.setState({
      journals: this.props.allJournal
        .slice(
          currentPage * NUMBER_PER_PAGE_JOURNAL,
          (currentPage * NUMBER_PER_PAGE_JOURNAL) + NUMBER_PER_PAGE_JOURNAL
        )
    });
  }

  /** Rendering jsx for the component.
    * @return {jsx} Return the jsx for the component.
  */
  render() {
    return (
      <UnauthenticatedUserCheck>
        <div className={`${styles.mainContentDiv}`}>
          <div className={`${styles.title}`}>Attached journal list</div>
          <div className="text-mute">(Not include <Link to={{ pathname: '/unattachedJournals', search: '?readingName=Unattached Journals' }}>unattached journals</Link>)</div>
          <LoadingAnimation />
          {/* Using the first reading id from the pingPongStates object. */}
          {this.state.journals.map(journal => (
            <JournalRow
              key={journal._id}
              journal={journal}
              readingId={Object.keys(journal.pingPongStates)[0]}
              isAllJournal
            />))}

          {this.props.allJournal.length !== 0 && <Pagination
            amount={this.props.allJournal.length}
            fetchContent={this.changeContent}
            numberPerpage={NUMBER_PER_PAGE_JOURNAL}
          />}
        </div>
      </UnauthenticatedUserCheck>
    );
  }
}

const mapStateToProps = state => ({
  user: state.user,
  allJournal: state.allJournal
});
const mapDispatchToProps = dispatch => ({
  fetchAllJournal: _ => dispatch(fetchAllJournal()),
  checkAuthentication: _ => dispatch(checkAuthentication())
});
export default connect(mapStateToProps, mapDispatchToProps)(AllJournalListContainer);
