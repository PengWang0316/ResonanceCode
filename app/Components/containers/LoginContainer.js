import React, { Component } from 'react';
import { connect } from 'react-redux';
import jquery from 'jquery';
import PropTypes from 'prop-types';

import styles from '../../styles/LoginContainer.module.css';
import LoginCoverImage from '../LoginCoverImage';
import LoginForm from '../LoginForm';
import { checkAuthentication } from '../../actions/UserActions';

/** The container for the login page. */
export class LoginContainer extends Component {
  static propTypes = {
    user: PropTypes.object.isRequired,
    checkAuthentication: PropTypes.func.isRequired
  };
  /** Checking the authentication when a user trys to load this page.
    * @return {null} No return.
  */
  componentWillMount() {
    // console.log(this.props.location.search);
    this.props.checkAuthentication(this.props.location.search); // Checking the authentication
  }

  /** Forwarding the page to reading after a user login successfully.
    * Showing the warning message when a user fails to login.
    * @param {object} nextProps is an object has new props that this component will receive.
    * @return {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    /* Pushing to the reading page after login */
    /* istanbul ignore next */
    if (nextProps.user.isAuth) this.props.history.push('/reading');
    else if (nextProps.user.loginErr) {
      jquery('#loginWarnMessage').css('opacity', '1');
      setTimeout(_ => { jquery('#loginWarnMessage').css('opacity', '0'); }, 3000);
    }
  }

  /** Render method for the component.
    * @return {jsx} Return the jsx for the component.
  */
  render() {
    return (
      <div className={`${styles.loginBackgroundContainer}`}>
        <LoginForm />
        <LoginCoverImage />
      </div>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({ user: state.user });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  checkAuthentication: jwt => dispatch(checkAuthentication(jwt))
});
export default connect(mapStateToProps, mapDispatchToProps)(LoginContainer);
