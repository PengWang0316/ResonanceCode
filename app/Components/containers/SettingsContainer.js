import React, { Component } from 'react';
import { connect } from 'react-redux';
import jQuery from 'jquery';

import UnauthenticatedUserCheck from '../SharedComponents/UnauthenticatedUserCheck';
import { checkAuthentication, updateSettingCoinMode } from '../../actions/UserActions';

/** Setting component */
class SettingsContainer extends Component {
  state = {
    // coinMode: 'btn btn-sm btn-secondary',
    // hexagramMode: 'btn btn-sm btn-outline-secondary'
    coinMode: ''
  };
  /** Checking the user authentication if the user's authentication is still false.
    * @returns {null} No return.
  */
  componentWillMount() {
    if (!this.props.user.isAuth) this.props.checkAuthentication();
  }

  /** Setting the initial state when the component receives the user object.
    * @param {object} nextProps is an object that contains props information.
    * @returns {null} No return.
  */
  componentWillReceiveProps(nextProps) {
    if (this.state.coinMode === '' && nextProps.user.isAuth) {
      // this.hexagramCoinMode = nextProps.user.settings.coinMode;
      this.setState({ coinMode: nextProps.user.settings.coinMode });
      // this.setCoinModeState();
    }
  }

  /** Setting coinMode state for the component.
    * @param {boolean} coinMode indicate whether a user click the coin mode.
    * @returns {null} No return.
  */
  setCoinModeState(coinMode) {
    this.setState({ coinMode });
    this.props.updateSettingCoinMode(coinMode);
    jQuery('#savedAlert').css({ opacity: 1 });
    if (this.savedTimeOut) clearTimeout(this.savedTimeOut);
    this.savedTimeOut = setTimeout(_ => jQuery('#savedAlert').css({ opacity: 0 }), 2000);
  }

  /** Changing the hexagram choose mode to coinMode.
    * @returns {null} No return.
  */
  handleCoinModeClick = () => {
    if (!this.state.coinMode) this.setCoinModeState(true);
  }

  /** Changing the hexagram choose mode to hexagramMode.
    * @returns {null} No return.
  */
  handleHexagramModeClick = () => {
    if (this.state.coinMode) this.setCoinModeState(false);
  }

  /** The render method for the component.
    * @returns {jsx} Return the jsx for the component.
  */
  render() {
    return (
      <UnauthenticatedUserCheck>
        <div className="container">
          <div className="font-weight-bold mb-4"><h5>Default mode to record hexagram</h5></div>

          <div className="btn-group" role="group" aria-label="Hexagram choose mode">
            <button type="button" className={this.state.coinMode ? 'btn btn-sm btn-secondary' : 'btn btn-sm btn-outline-secondary'} onClick={this.handleCoinModeClick}>Coin Mode</button>
            <button type="button" className={!this.state.coinMode ? 'btn btn-sm btn-secondary' : 'btn btn-sm btn-outline-secondary'} onClick={this.handleHexagramModeClick}>Line Mode</button>
          </div>

          <div id="savedAlert" className="alert alert-success saved-alert mt-3" role="alert">
            Saved Successfully!
          </div>

        </div>
      </UnauthenticatedUserCheck>
    );
  }
}
const mapStateToProps = state => ({
  user: state.user
});
const mapDispatchToProps = dispatch => ({
  checkAuthentication: _ => dispatch(checkAuthentication()),
  updateSettingCoinMode: coinMode => dispatch(updateSettingCoinMode(coinMode))
});
export default connect(mapStateToProps, mapDispatchToProps)(SettingsContainer);
