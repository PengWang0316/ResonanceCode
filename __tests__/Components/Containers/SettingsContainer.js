import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { SettingsContainer } from '../../../app/Components/containers/SettingsContainer';

jest.mock('jquery', () => jest.fn());
jest.mock('../../../app/Components/AlertPanel', () => 'AlertPanel');
jest.mock('../../../app/Components/UserGroups', () => 'UserGroups');
jest.mock('../../../app/Components/SharedComponents/UnauthenticatedUserCheck', () => 'UnauthenticatedUserCheck');

describe('SettingsContainer test', () => {
  const defaultProps = {
    user: { isAuth: true, settings: { coinMode: false, customName: 'customName' } },
    checkAuthentication: jest.fn(),
    updateSettingCoinMode: jest.fn(),
    saveCustomName: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<SettingsContainer {...props} />);

  test('constructor the user has coinMode, customName and authorized', () => {
    const component = getShallowComponent();
    const {
      coinMode, isShowAlert, customName, isNameChanged
    } = component.instance().state;
    expect(coinMode).toBe(false);
    expect(isShowAlert).toBe(false);
    expect(customName).toBe('customName');
    expect(isNameChanged).toBe(false);
    expect(component.instance().oldName).toBe('customName');
    expect(defaultProps.checkAuthentication).not.toHaveBeenCalled();
  });

  test('constructor the user has no coinMode, customName and unauthorized', () => {
    const component = getShallowComponent({ ...defaultProps, user: { isAuth: false, displayName: 'displayName' } });
    const {
      coinMode, isShowAlert, customName, isNameChanged
    } = component.instance().state;
    expect(coinMode).toBe('');
    expect(isShowAlert).toBe(false);
    expect(customName).toBe('displayName');
    expect(isNameChanged).toBe(false);
    expect(component.instance().oldName).toBe('displayName');
    expect(defaultProps.checkAuthentication).toHaveBeenCalledTimes(1);
  });

  test('componentWillReceiveProps has customName', () => {
    const component = getShallowComponent();
    component.setProps({ user: { isAuth: true, settings: { coinMode: true, customName: 'new name' } } });
    expect(component.instance().state.coinMode).toBe(true);
    expect(component.instance().state.customName).toBe('new name');
    expect(component.instance().oldName).toBe('new name');
  });

  test('componentWillReceiveProps has no customName', () => {
    const component = getShallowComponent();
    component.setProps({ user: { isAuth: true, displayName: 'display name', settings: { coinMode: false } } });
    expect(component.instance().state.coinMode).toBe(false);
    expect(component.instance().state.customName).toBe('display name');
    expect(component.instance().oldName).toBe('display name');
  });

  test('handleCoinModeClick coinMode is true', () => {
    const mockUpdateSettingCoinModeFn = jest.fn();
    const component = getShallowComponent({
      ...defaultProps, updateSettingCoinMode: mockUpdateSettingCoinModeFn
    });
    component.setState({ coinMode: true });
    component.find('button').at(1).simulate('click');
    expect(mockUpdateSettingCoinModeFn).not.toHaveBeenCalled();
  });

  test('handleCoinModeClick coinMode is false has savedTimeOut', () => {
    jest.useFakeTimers();
    const jquery = require('jquery');
    const mockCssFn = jest.fn();
    jquery.mockReturnValue({ css: mockCssFn });
    const mockUpdateSettingCoinModeFn = jest.fn();
    const component = getShallowComponent({
      ...defaultProps, updateSettingCoinMode: mockUpdateSettingCoinModeFn
    });
    component.instance().savedTimeOut = 1;
    component.instance().cancleAlertTimeOut = 2;
    component.setState({ coinMode: false });
    component.find('button').at(1).simulate('click');
    expect(mockUpdateSettingCoinModeFn).toHaveBeenLastCalledWith(true);
    expect(component.instance().state.coinMode).toBe(true);
    expect(component.instance().state.isShowAlert).toBe(true);
    expect(clearTimeout).toHaveBeenLastCalledWith(1);
    // expect(jquery).not.toHaveBeenCalled();
    // expect(mockCssFn).not.toHaveBeenCalled();
    jest.runAllTimers();
    expect(jquery).toHaveBeenCalled();
    expect(mockCssFn).toHaveBeenCalled();
    expect(component.instance().state.isShowAlert).toBe(false);
    expect(clearTimeout).toHaveBeenLastCalledWith(2);
  });

  test('handleHexagramModeClick coinMode is false', () => {
    const mockUpdateSettingCoinModeFn = jest.fn();
    const component = getShallowComponent({
      ...defaultProps, updateSettingCoinMode: mockUpdateSettingCoinModeFn
    });
    component.setState({ coinMode: false });
    component.find('button').at(2).simulate('click');
    expect(mockUpdateSettingCoinModeFn).not.toHaveBeenCalled();
  });

  test('handleHexagramModeClick coinMode is true has no savedTimeOut', () => {
    jest.useFakeTimers();
    const jquery = require('jquery');
    const mockCssFn = jest.fn();
    jquery.mockReturnValue({ css: mockCssFn });
    const mockUpdateSettingCoinModeFn = jest.fn();
    const component = getShallowComponent({
      ...defaultProps, updateSettingCoinMode: mockUpdateSettingCoinModeFn
    });
    // component.instance().savedTimeOut = 1;
    component.setState({ coinMode: true });
    component.find('button').at(2).simulate('click');
    expect(mockUpdateSettingCoinModeFn).toHaveBeenLastCalledWith(false);
    expect(component.instance().state.coinMode).toBe(false);
    expect(component.instance().state.isShowAlert).toBe(true);
    expect(clearTimeout).not.toHaveBeenCalled();
    // expect(jquery).not.toHaveBeenCalled();
    // expect(mockCssFn).not.toHaveBeenCalled();
    jest.runAllTimers();
    expect(jquery).toHaveBeenCalled();
    expect(mockCssFn).toHaveBeenCalled();
    expect(component.instance().state.isShowAlert).toBe(false);
    expect(clearTimeout).not.toHaveBeenCalled();
  });

  test('handleInputChange', () => {
    const component = getShallowComponent();
    component.find('input').simulate('change', { target: { value: '', id: 'id' } });
    expect(component.instance().state.isNameChanged).toBe(false);
    expect(component.instance().state.id).toBe('');
    component.instance().oldName = 'oldName';
    component.find('input').simulate('change', { target: { value: 'oldName', id: 'id' } });
    expect(component.instance().state.isNameChanged).toBe(false);
    expect(component.instance().state.id).toBe('oldName');
    component.find('input').simulate('change', { target: { value: 'A long long long long name', id: 'id' } });
    expect(component.instance().state.isNameChanged).toBe(false);
    expect(component.instance().state.id).toBe('A long long long long name');
    component.find('input').simulate('change', { target: { value: 'a correct name', id: 'id' } });
    expect(component.instance().state.isNameChanged).toBe(true);
    expect(component.instance().state.id).toBe('a correct name');
  });

  test('handleSaveNameBtnClick', () => {
    jest.useFakeTimers();
    const component = getShallowComponent();
    component.setState({ customName: 'newName' });
    component.find('button').at(0).simulate('click');
    expect(defaultProps.saveCustomName).toHaveBeenLastCalledWith('newName');
    expect(component.instance().oldName).toBe('newName');
    expect(component.instance().state.isNameChanged).toBe(false);
  });

  test('SettingsContainer snapshot', () => expect(renderer.create(<SettingsContainer {...defaultProps} />).toJSON()).toMatchSnapshot());
});
