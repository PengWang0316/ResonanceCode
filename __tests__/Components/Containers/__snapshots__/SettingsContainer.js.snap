// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsContainer test SettingsContainer snapshot 1`] = `
<UnauthenticatedUserCheck>
  <div
    className="container"
  >
    <div
      className="font-weight-bold h5 mb-2"
    >
      Setup your display name
    </div>
    <div
      className="text-muted mb-3"
    >
      <small>
        Other users can see it when you share readings to them. (No more than 20 characters.)
      </small>
    </div>
    <div
      className="form-group row mb-5"
      style={
        Object {
          "maxWidth": "470px",
        }
      }
    >
      <label
        className="col-sm-2 col-form-label"
        htmlFor="customName"
      >
        Name
      </label>
      <div
        className="col-sm-9"
      >
        <input
          className="form-control"
          id="customName"
          onChange={[Function]}
          placeholder="Display name..."
          type="text"
          value="customName"
        />
      </div>
      <div
        className="col-sm-1 mt-1 text-right"
      >
        <button
          className="btn btn-primary btn-sm"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Save
        </button>
      </div>
    </div>
    <div
      className="font-weight-bold h5 mb-3"
    >
      Default mode to record hexagram
    </div>
    <div
      className="d-flex flex-wrap align-items-center mb-4"
    >
      <div
        aria-label="Hexagram choose mode"
        className="btn-group mr-4"
        role="group"
      >
        <button
          className="btn btn-sm btn-outline-secondary"
          onClick={[Function]}
          type="button"
        >
          Coin Mode
        </button>
        <button
          className="btn btn-sm btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Line Mode
        </button>
      </div>
    </div>
    <div
      className="font-weight-bold h5 mt-5"
    >
      Manage your user groups
    </div>
    <div
      className="text-muted mb-3"
    >
      <small>
        You can share your readings with all group people after you setup your own group.
      </small>
    </div>
    <UserGroups
      userGroups={undefined}
    />
  </div>
</UnauthenticatedUserCheck>
`;
